src/main/java/com/impactupgrade/nucleus/service/segment/XeroDataSyncService.java

```diff
@@ -0,0 +1,81 @@
+package com.impactupgrade.nucleus.service.segment;
+
+import com.impactupgrade.nucleus.environment.Environment;
+import com.impactupgrade.nucleus.model.AccountingTransaction;
+import com.impactupgrade.nucleus.model.CrmContact;
+import com.impactupgrade.nucleus.model.CrmDonation;
+import com.impactupgrade.nucleus.model.PagedResults;
+
+import java.util.Calendar;
+import java.util.List;
+
+public class XeroDataSyncService implements DataSyncService {
+
+  protected Environment env;
+
+  @Override
+  public String name() {
+    return "xeroDataSync";
+  }
+
+  @Override
+  public boolean isConfigured(Environment env) {
+    return true;
+  }
+
+  @Override
+  public void init(Environment env) {
+    this.env = env;
+  }
+
+  // The method can/should be moved to an abstract class
+  // However, it depends on whether downstream platform support bulk contact processing or not -
+  // may be bulk vs 1 by 1 processing
+  //TODO: move to abstract class?
+  @Override
+  public void syncContacts(Calendar updatedAfter) throws Exception {
+    PagedResults<CrmContact> contactPagedResults = env.primaryCrmService().getDonorContacts(updatedAfter);
+    for (PagedResults.ResultSet<CrmContact> resultSet : contactPagedResults.getResultSets()) {
+      if (env.accountingPlatformService().isPresent()) {
+        try {
+          env.accountingPlatformService().get().updateOrCreateContacts(resultSet.getRecords());
+        } catch (Exception e) {
+          //TODO: process errors
+        }
+      }
+    }
+  }
+
+  @Override
+  public void syncTransactions(Calendar updatedAfter) throws Exception {
+    List<CrmDonation> crmDonations = env.primaryCrmService().getDonations(updatedAfter);
+    if (env.accountingPlatformService().isPresent()) {
+      //TODO: get all contacts ids for crm contacts and do bulk update instead?
```

Comment:
Safe to skip the updateOrCreateContacts step here! My goal is to schedule the contact sync first, then follow up with the transaction sync later, separated by at least an hour. Assume the contacts already exist and have been updated?

Reply:
Good point.
We may have one missing piece here: we have crmDonation and crmContact, but we need to have contact id (xero one) to assign transaction to contact.
It seems to me that we need to have smth like getContactForCrmContact method in the Xero service to complete the flow. What do you think?

Reply:
Really good point! That was one benefit of doing the real time sync, where we had everything about the contact and transaction in a single shot. Makes sense!

Reply:
Will implement the following flow then:
- Grab CrmContacts (already have them in CrmDonations)
- Find Xero Contacts for CrmContacts 
- Map Contacts By Name
- Add Xero contacts ids to transactions 
- Bulk update of Transactions

Reply:
Especially if the full names will now include the SFDC IDs, that by-name matching should be a lot easier...